### PR TITLE
Use exactly 2.3.3's standard library

### PR DESCRIPTION
**Pull request by chrisseaton**
I'm not sure how anyone's going to review this. I'm also surprised at what a huge diff we had from using the JRuby library. This is now simply MRI's `lib` from 2.3.3, a tiny bit of `ext`, and the whole of `lib` from the original `flori/json`. In some cases MRI has things we don't want (like `securerandom` because we have the pure Ruby RubySL version), so I've just put what we want first on the load path. We can't seem to run MRI's `webrick`. It wants socket methods we don't have, so I've stuck with the RubySL one for now.

### CODE REVIEW

File: doc/contributor/updating-ruby.md [hunk 1]
```diff
@@ -4,10 +4,12 @@ First, check with Chris Seaton for clearance.
 
 From MRI copy and paste over our versions of:
 
-* `lib/mri`
+* `lib` to `lib/mri`
+* `ext/{bigdecimal,json,psych,pty}/lib` to `lib/ext`
 * `test/mri`
 * `doc/legal/ruby-bsdl.txt` and `doc/legal/ruby-licence.txt`
 * `-h` and `--help` output in `CommandLineParser`
+* `lib/json` using the version of `flori/json` specified but not totally included in `ext/json`
```
Comment (eregon) [inline doc/contributor/updating-ruby.md]: It would be great to have a script to do this. I think JRuby had one in `tool`.

**Comment by eregon** [comment]
Why are some files in `lib/ext`? Files under `ext` in MRI just mean there is a native extension, but we don't import that so I'd rather have everything under `lib/mri` (or maybe `lib/ruby`).

**Comment by nirvdrum** [comment]
Don't forget to bump our version string afterwards.

**Comment by chrisseaton** [comment]
> Why are some files in `lib/ext`?

So that `lib/mri` is nothing more or less than exactly what is in MRI's `lib`. I think this will make working out what has changed easier.

**Comment by chrisseaton** [comment]
> Don't forget to bump our version string afterwards.

That's in this PR.

**Comment by chrisseaton** [comment]
@eregon two commits - one updates to 2.3.3 (I couldn't find a version that actually matched what we had - there were lots of differences against each, so 2.3.3 is as good as any) - the other is a script that is idempotent when run against 2.3.3 and can be run against any other version as well.

**Comment by eregon** [comment]
First of all, it's interesting to compare with an installed MRI 2.3.3, by looking at
`lib/ruby/2.3.0` and `lib/ruby/gems/2.3.0/gems`.

I think we should unify lib/mri and lib/ext, as it's also merged in a MRI distribution.

The diff with `git log -M` looks good overall. A few notes:
* `debug.rb` will likely not work (it's using continuations), do we want to include it anyway?
* `minitest` is removed from stdlib. In MRI it's a bundled gem (+MRI tests also use their own vendored copy of minitest).
* Truffle changes in `lib/mri/mkmf.rb` are lost, we need a way to reapply them somehow.
* `lib/mri/net/telnet.rb` was removed. In MRI it's a bundled gem.
* `lib/mri/racc*` except `racc/parser.rb` is gone, that's as in MRI 2.3.3 :+1: (only the runtime, not the full gem). @pitr-ch The debug puts with racc backtrace seem to be gone :)
* `lib/mri/racc/rdoc/grammar.en.rdoc` is extra, it's not in MRI distribution.
* `rake` is gone, it's a bundled gem in MRI.
* `lib/mri/resolv-replace.rb` is added (seems fine, it's to use `resolv.rb`, a pure-Ruby hostname resolving library)
* `lib/mri/rinda` is added, not sure why JRuby stripped it.
* `lib/mri/rss.rb` is added, not sure why JRuby stripped it.
* RubyGems was *downgraded* from 2.6.8 to VERSION = '2.5.2', hopefully this will not cause problems.
* `lib/mri/rubygems/ssl_certs/` was removed, not sure if it's needed.
* `lib/mri/securerandom.rb` added, I think we have a rubysl of this too.
* `lib/mri/shell` added, not sure why JRuby stripped it.
* `lib/mri/sync.rb` added
* `lib/mri/tempfile.rb` added (also rubysl)
* `lib/mri/test/unit.rb` removed, bundled gem in MRI
* `lib/mri/thwait.rb` added (seems good)
* `lib/mri/timeout.rb` added (also rubysl) I think the rubysl version is better as it does not create a Thread for every call to Timeout.timeout.
* `lib/mri/tracer.rb` looks like an interesting tool to test set_trace_func and TracePoint.
* `lib/mri/weakref.rb` added, for us it's core (a better impl). It's implemented in terms of ObjectSpace::WeakMap.new which we don't have. Maybe we should not import this since we have a completely different impl in core.
* `lib/mri/webrick.rb` added (also in rubysl)
* removed `lib/mri/win32`, seems we won't support that anytime soon. I think it's only installed on Windows.
* various MRI test updates: look OK

Should we import stdlib if we already have it in rubysl? It seems useless and potential for bugs, but it's simpler.

File: tool/update-mri.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+cp -r ../ruby/ext/bigdecimal/lib/bigdecimal lib/ext
+cp -r ../ruby/ext/psych/lib/psych lib/ext
+cp -r ../ruby/ext/psych/lib/*.rb lib/ext
+cp -r ../ruby/ext/pty/lib/*.rb lib/ext
+
+mv test/mri/excludes_truffle .
+rm -rf test/mri
+cp -r ../ruby/test test/mri
+mv excludes_truffle test/mri
+rm -rf test/mri/excludes
```
Comment (eregon) [inline tool/update-mri.sh]: You could also use the `git checout` trick on test/mri/excludes_truffle

**Comment by eregon** [comment]
Ci errors show missing rake and updated MRI tests.

**Comment by chrisseaton** [comment]
We're going to put this on hold while @bjfish extends his previous work on bundled gems to bundle all the others which are confusing this update.

**Comment by bjfish** [comment]
I've merged the bundled gem PR: https://github.com/graalvm/truffleruby/pull/52

I'll review the default specs after the stdlib update happens.

**Comment by chrisseaton** [comment]
I removed code duplicated in RubySL, added a note on updating `mkmf` and reduced the diff there.

**Comment by eregon** [comment]
@chrisseaton What about our discussion about merging lib/ext and lib/mri like in MRI distributions?

**Comment by chrisseaton** [comment]
Sorry do you still want to merge `ext` and `mri`?

What do you think the advantage is? Less on the load path? I don't really mind.

**Comment by eregon** [comment]
Yes, I think they both are part of stdlib and all Ruby files of stdlib should be in one directory.

**Comment by chrisseaton** [comment]
See #57

**Pull request merged**